### PR TITLE
fix(foreach): Remove shell invocation from foreach

### PR DESCRIPTION
**Pull request by annettejanewilson**
As discussed in [issue 135], turbolift applies shell escaping inconsistently on the arguments in foreach. Upon consideration, I think there is no good reason to apply shell escaping at all. It makes turbolift more fragile, more confusing and ultimately less flexible.

Instead, do not use a shell at all and all these problems disappear. None of the examples in the README use this functionality. If users do need to invoke a shell, for example to do redirection, it is much more explicit and understandable for them if they do it themselves.

This PR:

* Removes the use of a shell, instead directly executing a subprocess based on the arguments provided.
* Removes custom handling of arguments. Users are directed to the standard GNU-style `--` to indicate that everything else on the command-line is an argument, not an option. This makes the code substantially simpler.
* Uses shellescape to format the printed command. (In retrospect I could separate this into another PR. Let me know if you prefer that.)

[issue 135]: https://github.com/Skyscanner/turbolift/issues/135

### CODE REVIEW

File: cmd/foreach/foreach.go
Comment (annettejanewilson) [inline cmd/foreach/foreach.go:37]: Ah, I think I need to reformat this.

File: cmd/foreach/foreach.go
Comment (annettejanewilson) [inline cmd/foreach/foreach.go:65]: I may have removed again everything that required me to switch to runE. This can probably be simplified.

**Review by morwoen — APPROVED** [review]

**Comment by rnorth** [comment]
Let's think about this, as it does look nicer - FYI @sledigabel 

I need to backtrack and check why we used a shell in the first place. From rough memory I vaguely recall frustration with there being enough real-world situations where a shell was needed, but that could be mistaken. I'll check slack history, particularly back on the internal pre-go (bash!) prototype of turbolift.

**Comment by rnorth** [comment]
@sledigabel, @Dan7-7-7 and I had our regular scheduled catchup yesterday; we're kinda on the fence about this, partly because it would be a breaking change and requires ordinary users (who are not necessarily POSIX power users) to learn the `--` option. We're not so fussed about the volume of code we have to maintain.

We also discussed, and looked up in Slack, the scenarios where shell has been desirable in the past; these are roughly in two categories:

* compatibility with SDK version/environment managers. For me this is the number one priority, since tools like ASDF can drastically simplify working across many repos that each has its own version/environment specifics
* intentional use of chained commands - e.g. patterns like `turbolift foreach 'gh pr view --comments | grep "Closing temporarily" && gh pr reopen --comment "Reopening"'`. *I have a bad feeling we've already broken this in a recent release*

Running some quick local tests with this branch I think we're safe-ish on the first but it would be a breaking change for the second:

```
> turbolift foreach -- which java
  OK   Reading campaign data (repos.txt)
  OK   Executing which java in work/someorg/somerepo

     Executing: which [java] in work/someorg/somerepo
     /Users/richardnorth/.asdf/shims/java
  OK   turbolift foreach completed (1 OK, 0 skipped)
```
^ Evidently the ASDF shims are still used

```
> cd work/someorg/somerepo
someorg/somerepo> asdf local java temurin-21.0.0+35.0.LTS
someorg/somerepo> cd -
> turbolift foreach -- java --version
  OK   Reading campaign data (repos.txt)
  OK   Executing java --version in work/someorg/somerepo

     Executing: java [--version] in work/someorg/somerepo
     openjdk 21 2023-09-19 LTS
     OpenJDK Runtime Environment Temurin-21+35 (build 21+35-LTS)
     OpenJDK 64-Bit Server VM Temurin-21+35 (build 21+35-LTS, mixed mode)
  OK   turbolift foreach completed (1 OK, 0 skipped)
```
^ ... and actually apply

```
> turbolift foreach -- which java && echo hello
  OK   Reading campaign data (repos.txt)
  OK   Executing which java in work/someorg/somerepo

     Executing: which [java] in work/someorg/somerepo
     /Users/richardnorth/.asdf/shims/java
  OK   turbolift foreach completed (1 OK, 0 skipped)

hello
```
^ using shell syntax behaves as might be expected. *Not shown:* wrapping a scriptlet in single quotes used to work, but now doesn't work in our release version.

```
> turbolift foreach -- sh -c "which java && echo hello"
  OK   Reading campaign data (repos.txt)
  OK   Executing sh -c 'which java && echo hello' in work/someorg/somerepo

     Executing: sh [-c which java && echo hello] in work/someorg/somerepo
     /Users/richardnorth/.asdf/shims/java
     hello
  OK   turbolift foreach completed (1 OK, 0 skipped)
```
^ workaround necessary to execute multiple commands, if one does not want to put them in an actual script file.

**Comment by rnorth** [comment]
One other thing we wondered: given now mandatory use of `--`, is it at least possible to make the command fail with a sensible and helpful error message if the user does not supply it?

**Comment by annettejanewilson** [comment]
FYI I fully implemented this with a `-c/--command` option that would take a shell command as an option, then realised that it had very little value. It was just sugar that turned `turbolift foreach sh -c 'command | with | pipes'` into `turbolift foreach -c 'command | with | pipes'` but left you wondering about the details.

My feeling is that it's best to provide a simple and predictable building block that handles 90% of cases as-is and which doesn't get in the way when you want to do the complex ones: you can easily invoke a shell yourself because you can be confident that your arguments get through to it untouched.

This will need release notes, but I also know people are likely to be using this through brew and not seeing release notes. Do you think there's anything needed for the change in behaviour?

I see 2.4.1 was released yesterday. I was hoping we could resolve this before releasing that.

Here's a handy summary:

<table>
<tr><th>What works?</th><th>2.4.0</th><th>2.4.1</th><th>This branch</th></tr>
<tr><td>Can pass any valid shell script as a single argument</td><td>Yes</td><td>No</td><td>No</td></tr>
<tr><td>Can pass simple command directly as args (no empty args, no spaces, no shell chars)</td><td>Yes</td><td>Yes</td><td>Yes</td></tr>
<tr><td>Can pass simple commands with spaces directly as args (no empty args, no shell chars)</td><td>No</td><td>Yes</td><td>Yes</td></tr>
<tr><td>Can pass <em>any</em> command directly as args (including empty args, arbitrary chars in args)</td><td>No</td><td>No</td><td>Yes</td></tr>
<tr><td>There is <em>some</em> way to pass an arbitrary shell script</td><td>Yes (single arg)</td><td>No</td><td>Yes (<code>sh -c ...</code>)</td></tr>
</table>

**Comment by annettejanewilson** [comment]
> One other thing we wondered: given now mandatory use of `--`, is it at least possible to make the command fail with a sensible and helpful error message if the user does not supply it?

Not really. It's not _strictly_ mandatory. It's just got the normal GNU-style processing. I just think it's a lot 

> One other thing we wondered: given now mandatory use of `--`, is it at least possible to make the command fail with a sensible and helpful error message if the user does not supply it?

We could check that `c.ArgsLenAtDash()` is `0`. It's `-1` if it doesn't appear at all, and positive if some number of non-option arguments appear before it.

Note that it's not _mandatory_, just strongly recommended to avoid confusion.

**Comment by sledigabel** [comment]
> Note that it's not _mandatory_, just strongly recommended to avoid confusion.

I think in the current state of the PR it is mandatory in order to avoid the command options from being swallowed by Cobra since we're not disabling the parsing any longer. So it's either using `--` or quoting the command.

Cobra is too clever for this so checking the arg length sounds like a good strategy.

**Comment by rnorth** [comment]
Yes, it can be avoided but quickly becomes needed if any flags are used for the command.

I was going to say we could emit a prominent warning, but most people when faced with a warning and an error only see the error.

Perhaps we should just make it mandatory, and bomb out with our own very clear message if it's not been used?

**Comment by annettejanewilson** [comment]
I've pushed a change to make `--` mandatory. However, there are some caveats:

* When you omit `--` and include no options, you'll see "Error: Use -- to separate command" and brief usage instructions.
* When you omit `--` and _do_ include options in your command, you'll see something like "Error: unknown shorthand flag: 'x' in -x" and brief usage instructions.
* When you specify `--help`, you will see detailed usage instructions. These are very clear about the need for `--`.

The problem here is that the brief usage instructions shown when the arguments are incorrect do not include either the "Long" _or_ the "Short" description of the command. See [cobra #1084].

[cobra #1084]: https://github.com/spf13/cobra/issues/1084

**Comment by annettejanewilson** [comment]
Oh, lol, now I type that out I see that `[flags]` floating at the end. Let me fix that.

**Comment by annettejanewilson** [comment]
Oh, I can't! It's put there by Cobra. Huh. Why does it put it at the _end_? Who puts flags at the end?

**Comment by annettejanewilson** [comment]
Okay, I'm happy with this again now. I've rebased again `main` (cleanly) but otherwise I haven't collapsed or rewritten any of my commits.

**Comment by annettejanewilson** [comment]
Please let me know if you would prefer the invocation formatting in a separate PR. I think maybe that would make sense as it's not a necessary part of this fix, more a feature in itself, and it should probably be applied more broadly than just to foreach. I note that right now it seems like all the invocations get reported _twice_ by different parts of the code - once in foreach and again in executor. Again, that might be better dealt with in its own PR. I already did the work to separate it into its own commit, so I think it could easily be dropped from this PR.

**Comment by sledigabel** [comment]
I think it looks good @annettejanewilson. The formatting can stay in the same PR as far as I'm concerned, it's a nice addition to this refactor.

We can prob live with the error message triggered if the user omits the `--` and adds an option. I understand it's because Cobra parses it before you can handle this kind of error? I think it's OK because despite the error handling saying the flag doesn't exist, the usage makes it very clear what the user needs to do.

File: cmd/foreach/foreach.go
Comment (annettejanewilson) [reply cmd/foreach/foreach.go:65]: Ended up needing this again. We change to `runE` so that we can fail with an error if the user doesn't put `--` before their command.

**Review by sledigabel — DISMISSED** [review]
LGTM, great work @annettejanewilson thanks

File: README.md
Comment (Dan7-7-7) [inline README.md:143]: Sorry, a bit of a nitpick, and something we should have had already, but can we add an example using a script with the new requirement of `-- sh -c`? Possibly along with a note that you need to specify the full path or the path relative to the work/org/repo directory, since the command is executed in there (we've had questions about this before).
I suspect this won't be obvious to everyone.

**Comment by annettejanewilson** [comment]
Suggested text for release notes:
> Turbolift's `foreach` command no longer invokes a shell to interpret your command. This fixes some issues around passing an empty string as an argument and difficulty escaping shell characters. However, it also means that you can no longer directly pass turbolift a quoted script containing redirects or pipes. Instead, you will need to explicitly invoke a shell yourself. For example, if you previously used `turbolift foreach 'script | with > redirects'` you would now use `turbolift foreach -- bash -c 'script | with > redirects'`. We hope that the improved consistency and predictability makes up for needing to be a bit more explicit in this advanced usage.

File: README.md
Comment (annettejanewilson) [reply README.md:143]: Also, don't worry about nitpicking! I appreciate the feedback. :D

Comment (Dan7-7-7) [reply README.md:143]: thanks!

**Review by Dan7-7-7 — APPROVED** [review]
LGTM, the change might cause minor disruption for users but I think ultimately it's more consistent / less confusing this way. 
Re-approving after my comment but will let @rnorth and @sledigabel have the final say.

**Pull request merged**
